File: app/src/main/java/com/bhupendra/foodorderingapp/adapters/CategoryAdapter.java
package com.bhupendra.foodorderingapp.adapters;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.LinearGradient;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.Shader;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.text.Layout;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bhupendra.foodorderingapp.R;
import com.bhupendra.foodorderingapp.data.CategoryInfo;
import com.bhupendra.foodorderingapp.utils.FirebaseConnector;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.storage.StorageReference;
import com.squareup.picasso.Callback;
import com.squareup.picasso.Picasso;
import com.squareup.picasso.Target;

import java.util.ArrayList;

public class CategoryAdapter extends RecyclerView.Adapter<CategoryAdapter.CategoryHolder> {

    ArrayList<CategoryInfo> list;
    Context context;

    public CategoryAdapter(ArrayList<CategoryInfo> categoryInfo, Context context) {
        this.list = categoryInfo;
        this.context = context;
    }

    @NonNull
    @Override
    public CategoryHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new CategoryHolder(LayoutInflater.from(context).inflate(R.layout.recycler_category,parent,false));
    }

    @Override
    public void onBindViewHolder(@NonNull final CategoryHolder holder, int position) {
        CategoryInfo categoryInfo = list.get(position);

        holder.image.setVisibility(View.INVISIBLE);
        StorageReference load = FirebaseConnector.getStorageReference().child("menu").child(categoryInfo.getImageURL());

        final Target target = new Target() {
            @Override
            public void onBitmapLoaded(Bitmap bitmap, Picasso.LoadedFrom from) {
                Bitmap newBitmap = addGradient(bitmap);

                holder.image.setImageBitmap(newBitmap);
                holder.image.setVisibility(View.VISIBLE);
            }

            @Override
            public void onBitmapFailed(Exception e, Drawable errorDrawable) {
                holder.image.setVisibility(View.VISIBLE);
            }

            @Override
            public void onPrepareLoad(Drawable placeHolderDrawable) {
            }
        };

        load.getDownloadUrl().addOnSuccessListener(new OnSuccessListener<Uri>() {
            @Override
            public void onSuccess(Uri uri) {
                Log.e("DATA",uri.toString());

                Picasso.get().load(uri.toString()).placeholder(R.drawable.none).into(target);

            }
        }).addOnFailureListener(new OnFailureListener() {
            @Override
            public void onFailure(@NonNull Exception exception) {
                // Handle any errors
            }
        });
        holder.textTitle.setText(categoryInfo.getName());
    }

    public Bitmap addGradient(Bitmap originalBitmap) {
        int width = originalBitmap.getWidth();
        int height = originalBitmap.getHeight();
        Bitmap updatedBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(updatedBitmap);

        canvas.drawBitmap(originalBitmap, 0, 0, null);

        Paint paint = new Paint();
        LinearGradient shader = new LinearGradient(0, 0, width, 0, context.getResources().getColor(R.color.gradientendColor), context.getResources().getColor(R.color.gradientstartColor), Shader.TileMode.CLAMP);
        paint.setShader(shader);
        paint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_IN));
        canvas.drawRect(0, 0, width, height, paint);

        return updatedBitmap;
    }
    @Override
    public int getItemCount() {
        return list.size();
    }

    class CategoryHolder extends RecyclerView.ViewHolder
    {
        TextView textTitle;
        ImageView image;
        public CategoryHolder(View itemView) {
            super(itemView);
            textTitle = itemView.findViewById(R.id.textTitle);
            image = itemView.findViewById(R.id.image);
        }
    }
}

File: app/src/main/java/com/adityagupta/foodorderingapp/adapters/CategoryAdapter.kt
package com.adityagupta.foodorderingapp.adapters

import android.content.Context
import android.content.Intent
import android.graphics.*
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.adityagupta.foodorderingapp.R
import com.adityagupta.foodorderingapp.data.CategoryInfo
import kotlinx.android.synthetic.main.recycler_category.view.*
import android.graphics.drawable.Drawable
import android.graphics.Bitmap
import android.util.Log
import android.widget.ImageView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.target.Target
import android.graphics.drawable.BitmapDrawable
import com.adityagupta.foodorderingapp.activities.CategoryActivity
import com.adityagupta.foodorderingapp.data.CategoryItem
import com.bumptech.glide.request.RequestOptions
import java.util.*


class CategoryAdapter(private val list: ArrayList<CategoryInfo>, private val context: Context) : RecyclerView.Adapter<CategoryHolder>() {


    override fun onCreateViewHolder(p0: ViewGroup, p1: Int): CategoryHolder {
        return CategoryHolder(LayoutInflater.from(context).inflate(R.layout.recycler_category, p0, false))
    }

    override fun getItemCount(): Int {
        return list.size
    }

    inner class compMin : Comparator<CategoryItem> {
        override fun compare(a: CategoryItem, b: CategoryItem): Int {
            if (a.price > b.price)
                return -1 // highest value first
            return if (a.price === b.price) 0 else 1
        }
    }

    inner class compMax : Comparator<CategoryItem> {
        override fun compare(a: CategoryItem, b: CategoryItem): Int {
            if (a.price < b.price)
                return -1 // highest value first
            return if (a.price === b.price) 0 else 1
        }
    }

    override fun onBindViewHolder(holder: CategoryHolder, p1: Int) {
        val category: CategoryInfo = list[p1]

        holder.textTitle.text = category.name

        var min : CategoryItem = Collections.max(category.items, compMin())
        var max : CategoryItem = Collections.max(category.items, compMax())
        if( min != max)
            holder.textPriceRange.text = "₹${min.price} - ₹${max.price}"
        else
            holder.textPriceRange.text = "₹${min.price}"

        val req = RequestOptions()
        req.placeholder(R.drawable.none)
        req.error(R.drawable.none)

        Glide.with(context)
                .setDefaultRequestOptions(req)
                .load(category.imageURL).listener(object : RequestListener<Drawable> {
                    override fun onLoadFailed(e: GlideException?, model: Any?, target: Target<Drawable>?, isFirstResource: Boolean): Boolean {
                        Log.e("FailedGlide", "YES")
                        return false
                    }

                    override fun onResourceReady(resource: Drawable?, model: Any?, target: Target<Drawable>?, dataSource: DataSource?, isFirstResource: Boolean): Boolean {
                        Log.e("FailedGlide", "NO")

                        val bit: Bitmap = drawableToBitmap(resource!!)!!
                        val netBit = addGradient(bit)

                        holder.image.setImageBitmap(netBit)
                        holder.image.visibility = View.VISIBLE
                        return true
                    }

                }).into(holder.image)

        holder.itemView.setOnClickListener {
            val intent = Intent(context, CategoryActivity::class.java)
            intent.putExtra("index", p1)
            context.startActivity(intent)
        }
    }

    fun drawableToBitmap(drawable: Drawable): Bitmap? {
        var bitmap: Bitmap? = null

        if (drawable is BitmapDrawable) {
            if (drawable.bitmap != null) {
                return drawable.bitmap
            }
        }

        if (drawable.intrinsicWidth <= 0 || drawable.intrinsicHeight <= 0) {
            bitmap = Bitmap.createBitmap(1, 1, Bitmap.Config.ARGB_8888) // Single color bitmap will be created of 1x1 pixel
        } else {
            bitmap = Bitmap.createBitmap(drawable.intrinsicWidth, drawable.intrinsicHeight, Bitmap.Config.ARGB_8888)
        }

        val canvas = Canvas(bitmap)
        drawable.setBounds(0, 0, canvas.width, canvas.height)
        drawable.draw(canvas)
        return bitmap
    }

    fun addGradient(originalBitmap: Bitmap): Bitmap {
        val width = originalBitmap.width
        val height = originalBitmap.height
        val updatedBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(updatedBitmap)

        canvas.drawBitmap(originalBitmap, 0f, 0f, null)

        val paint = Paint()
        val shader = LinearGradient(0f, 0f, width.toFloat(), 0f, context.resources.getColor(R.color.gradientendColor), context.resources.getColor(R.color.gradientstartColor), Shader.TileMode.CLAMP)
        paint.shader = shader
        paint.xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_IN)
        canvas.drawRect(0f, 0f, width.toFloat(), height.toFloat(), paint)

        return updatedBitmap
    }

}

class CategoryHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
    val textTitle = itemView.textTitle!!
    val image: ImageView = itemView.image as ImageView
    val textPriceRange = itemView.pricerange
}